include/circt/Dialect/SMT/SMTOps.td

```diff
@@ -337,4 +338,64 @@ def ImpliesOp : SMTOp<"implies", [Pure]> {
   let assemblyFormat = "$lhs `,` $rhs attr-dict";
 }
 
+class QuantifierOp<string mnemonic> : SMTOp<mnemonic, [
+  RecursivelySpeculatable,
+  RecursiveMemoryEffects,
+  SingleBlockImplicitTerminator<"smt::YieldOp">,
+]> {
+  let description = [{
+    This operation represents the }] # summary # [{ as described in the
+    [SMT-LIB 2.6 standard](https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf).
+    It is part of the language itself rather than a theory or logic.
+
+    The operation specifies the names (as an attribute) and types (as the types
+    of the block arguments of the regions) of bound variables that may be
+    used in the 'body' of the operation. If a 'patterns' region is specified,
+    the block arguments must match the ones of the 'body' region and (other
+    than there) must be used at least once in the 'patterns' region. It may also
+    not contain any operations that bind variables, such as quantifiers. While
+    the 'body' region must always yield exactly one `!smt.bool`-typed value, the
+    'patterns' region can yield an arbitrary number (but at least one) of SMT
+    values.
+
+    The 'no_patterns' attribute is only allowed when no 'patterns' region is
+    specified and forbids the solver to generate and use patterns for this
+    quantifier.
+
+    The 'weight' attribute indicates the importance of this quantifier being
+    instantiated compared to other quantifiers that may be present. The default
+    value is zero.
+  }];
+
+  let arguments = (ins DefaultValuedAttr<I32Attr, "0">:$weight,
+                       UnitAttr:$noPattern,
+                       StrArrayAttr:$boundVarNames);
```

Comment:
Can we make this optional somehow?
I'm not sure we want to force users to give names?

Reply:
Yes, that makes sense.